src/goto-instrument/contracts/contracts.cpp

```diff
@@ -450,44 +489,82 @@ void code_contractst::check_apply_loop_contracts(
   {
     for(size_t i = 0; i < new_decreases_vars.size(); i++)
     {
-      code_assignt new_decreases_assignment{new_decreases_vars[i],
-                                            decreases_clause_exprs[i]};
-      new_decreases_assignment.add_source_location() =
-        loop_head->source_location();
-      converter.goto_convert(new_decreases_assignment, generated_code, mode);
+      code_assignt new_decreases_assignment{
+        new_decreases_vars[i], decreases_clause_exprs[i]};
+      new_decreases_assignment.add_source_location() = loop_head_location;
+      converter.goto_convert(
+        new_decreases_assignment, pre_loop_end_instrs, mode);
     }
 
     // Generate: assertion that the multidimensional decreases clause's value
-    // after the loop is smaller than the value before the loop.
-    // Here, we use the lexicographic order.
+    // after the loop is lexicographically smaller than its initial value.
     code_assertt monotonic_decreasing_assertion{
       generate_lexicographic_less_than_check(
         new_decreases_vars, old_decreases_vars)};
-    monotonic_decreasing_assertion.add_source_location() =
-      loop_head->source_location();
+    monotonic_decreasing_assertion.add_source_location() = loop_head_location;
     converter.goto_convert(
-      monotonic_decreasing_assertion, generated_code, mode);
-    generated_code.instructions.back().source_location_nonconst().set_comment(
-      "Check decreases clause on loop iteration");
+      monotonic_decreasing_assertion, pre_loop_end_instrs, mode);
+    pre_loop_end_instrs.instructions.back()
+      .source_location_nonconst()
+      .set_comment("Check decreases clause on loop iteration");
 
     // Discard the temporary variables that store decreases clause's value
     for(size_t i = 0; i < old_decreases_vars.size(); i++)
     {
-      generated_code.add(goto_programt::make_dead(
-        old_decreases_vars[i], loop_head->source_location()));
-      generated_code.add(goto_programt::make_dead(
-        new_decreases_vars[i], loop_head->source_location()));
+      pre_loop_end_instrs.add(
+        goto_programt::make_dead(old_decreases_vars[i], loop_head_location));
+      pre_loop_end_instrs.add(
+        goto_programt::make_dead(new_decreases_vars[i], loop_head_location));
     }
   }
 
   insert_before_swap_and_advance(
     goto_function.body,
     loop_end,
-    add_pragma_disable_assigns_check(generated_code));
+    add_pragma_disable_assigns_check(pre_loop_end_instrs));
 
   // change the back edge into assume(false) or assume(guard)
   loop_end->turn_into_assume();
   loop_end->set_condition(boolean_negate(loop_end->get_condition()));
+
+  std::set<goto_programt::targett> seen_targets;
+  // Find all exit points of the loop, make temporary variables `DEAD`,
+  // and check that step case was checked for non-vacuous loops.
+  for(const auto &t : loop)
+  {
+    if(!t->is_goto())
```

Comment:
Are all loop exit points necessarily GOTOs instructions?

Isn't it possible to exit the loop just by executing the instruction that comes right after an instruction that is in the loop in sequential order ?

Reply:
Could we try the condition 
"the instruction is a GOTO with a target that is not in the set of loop instructions OR the next instruction in sequential order is not in the set of loop instructions ?"

Reply:
Just because we have some instructions after the loop, doesn't mean they would be executed automatically, because the last instruction of a loop is a back edge GOTO. We must either jump out via a `break` or `return` or a `goto` on loop guard falsification at the loop head -- all 3 of these emit `GOTO` instructions.